### PR TITLE
Fix boost linking errors for Windows

### PR DESCRIPTION
**Pull request by Ace314159**
### Description

For Windows, #900 isn't enough to fix the boost linking errors and similar fixes were needed in `background_processing` and `profiler`. This also needs to be backported to foxy.

### Checklist
- [ ] **Required by CI**: Code is auto formatted using [clang-format](http://moveit.ros.org/documentation/contributing/code)
- [ ] Extend the tutorials / documentation [reference](http://moveit.ros.org/documentation/contributing/)
- [ ] Document API changes relevant to the user in the [MIGRATION.md](https://github.com/ros-planning/moveit/blob/master/MIGRATION.md) notes
- [ ] Create tests, which fail without this PR [reference](https://ros-planning.github.io/moveit_tutorials/doc/tests/tests_tutorial.html)
- [ ] Include a screenshot if changing a GUI
- [ ] While waiting for someone to review your request, please help review [another open pull request](https://github.com/ros-planning/moveit/pulls) to support the maintainers

[//]: # "You can expect a response from a maintainer within 7 days. If you haven't heard anything by then, feel free to ping the thread. Thank you!"

### CODE REVIEW

**Review by vatanaksoytezer — APPROVED** [review]

**Review by JafarAbdi — CHANGES_REQUESTED** [review]
We shouldn't need an explicit link against individual targets in Boost since we already have Boost in ament_target_dependencies

I'm on on my phone right now so I can't push the fix will do it later today

**Comment by Ace314159** [comment]
The issue seems to arise because when doing `find_package(geometric_shapes)` it calls `find_package(Boost ...)` again with different and fewer components than what `moveit_core` initially calls it with, causing `Boost_LIBRARIES` to be overwritten. Then, `ament_target_dependencies` can't properly link all the Boost components.

I wasn't sure how to prevent `Boost_LIBRARIES` from being overwritten, so I chose to manually link. I was able to find [this thread](https://cmake.org/pipermail/cmake/2009-December/034099.html) detailing a similar issue, but according to them, just maintaining the internal Boost variables should be sufficient to avoid `Boost_LIBRARIES`, but that doesn't seem to be the case anymore from what I observed.

If you were were able to find a better fix, I'd be happy to use that instead.

**Comment by JafarAbdi** [comment]
> The issue seems to arise because when doing `find_package(geometric_shapes)` it calls `find_package(Boost ...)` again with different and fewer components than what `moveit_core` initially calls it with, causing `Boost_LIBRARIES` to be overwritten. Then, `ament_target_dependencies` can't properly link all the Boost components.
> 
> I wasn't sure how to prevent `Boost_LIBRARIES` from being overwritten, so I chose to manually link. I was able to find [this thread](https://cmake.org/pipermail/cmake/2009-December/034099.html) detailing a similar issue, but according to them, just maintaining the internal Boost variables should be sufficient to avoid `Boost_LIBRARIES`, but that doesn't seem to be the case anymore from what I observed.
> 
> If you are able to find a better fix, I'd be happy to use that instead.

Thanks for the detailed explanation and sorry for the late response, we discussed this last moveit stand-up and we're not sure what the best solution is, we either need to 
1- run [include(ConfigExtras.cmake)](https://github.com/ros-planning/moveit2/blob/main/moveit_core/CMakeLists.txt#L13-L14) after calling all `find_packages(...)` and rely on ament_* to link (we need to make sure no call to `find_packages(...)` depend on Boost)
2- Remove Boost from `ament_target_dependencies` calls and manually specify each boost component like what's done in this PR but we have to do it throughout the whole codebase

Personally, I prefer the first solution since it takes advantage of the ament functionalities

**Comment by Ace314159** [comment]
The first solution sounds good to me too. I'll make the necessary changes.

File: moveit_core/CMakeLists.txt [hunk 1]
```diff
@@ -52,6 +49,9 @@ find_package(pluginlib REQUIRED)
 # TODO: Port python bindings
 # find_package(pybind11 REQUIRED)
 
+# Finds Boost Components
+include(ConfigExtras.cmake)
+
```
Comment (tylerjw) [inline moveit_core/CMakeLists.txt:54]: It looks like you just moved this line?

Comment (Ace314159) [reply moveit_core/CMakeLists.txt:54]: Sorry, I forgot to revert the previous commit as well. Is that what you were implying was missing?

Comment (tylerjw) [reply moveit_core/CMakeLists.txt:54]: Further up in this file you removed these two lines and then added them back here.  I'm confused about what difference that would make.

Comment (Ace314159) [reply moveit_core/CMakeLists.txt:54]: As per what Jafar said, the first solution involves moving the `include(ConfigExtras.cmake)` line after all the `find_pakage` calls. When we do this, the `include(ConfigExtras.cmake)` will be the last one to call `find_package(Boost ...)`, so `Boost_LIBRARIES` will not be overwritten by any dependencies, allowing us to use ament_* to link.

That's what I did when I removed those two lines from above and added them back there.

**Review by tylerjw — APPROVED** [review]

File: moveit_core/CMakeLists.txt (hunk 1 above)
Comment (tylerjw) [reply moveit_core/CMakeLists.txt:54]: Thank you for the explination.  I misunderstood what this would do.

**Comment by JafarAbdi** [comment]
@Ace314159 Super sorry for the delay, do you mind testing this PR ? I'll merge then immediately

File: moveit_ros/planning_interface/CMakeLists.txt [hunk 2]
```diff
@@ -79,7 +79,6 @@ add_subdirectory(move_group_interface)
 if(BUILD_TESTING)
   find_package(ament_cmake_gtest REQUIRED)
   find_package(ros_testing REQUIRED)
-  find_package(Boost REQUIRED COMPONENTS filesystem)
```
Comment (JafarAbdi) [inline moveit_ros/planning_interface/CMakeLists.txt:82]: This component is already in `ConfigExtras.cmake`

**Review by JafarAbdi — APPROVED** [review]

**Comment by vatanaksoytezer** [comment]
If we are reverting #900. We should also revert #925 as well.

**Comment by Ace314159** [comment]
@JafarAbdi This looks good on Windows.

**Review by tylerjw — APPROVED** [review]

**Pull request merged**